engine/api/src/main/java/io/deephaven/engine/table/updateBySpec/EmaSpec.java

```diff
@@ -1,46 +1,105 @@
  */
 @Immutable
-@SimpleStyle
+@BuildableStyle
 public abstract class EmaSpec implements UpdateBySpec {
+    // TimeScale is used internally to manage timed vs. tick EMA information
+    public static class TimeScale {
+        private final String timestampCol;
+        private final long timescaleUnits;
+
+        private TimeScale(String timestampCol, long timescaleUnits) {
+            this.timestampCol = timestampCol;
+            this.timescaleUnits = timescaleUnits;
+        }
+
+        public static TimeScale ofTime(@NotNull final String timestampCol, long timeScaleNanos) {
+            return new TimeScale(timestampCol, timeScaleNanos);
+        }
+
+        public static TimeScale ofTicks(long tickWindow) {
+            return new TimeScale(null, tickWindow);
+        }
+
+        public boolean isTimeBased() {
+            return timestampCol != null;
+        }
+
+        public String timestampCol() {
+            return timestampCol;
+        }
+
+        /**
+         * This returns the timescale units (ticks or nanos)
+         *
+         * @return the timescale units (ticks or nanos)
+         */
+        public long timescaleUnits() {
+            return timescaleUnits;
+        }
+    }
+
     public static EmaSpec ofTime(@NotNull final EmaControl control,
-            @NotNull final String timestampCol,
-            long timeScaleNanos) {
-        return ImmutableEmaSpec.of(control, timestampCol, timeScaleNanos);
+                                 @NotNull final String timestampCol,
+                                 long timeScaleNanos) {
+        return ImmutableEmaSpec.builder()
+                .control(control)
+                .timeScale(TimeScale.ofTime(timestampCol, timeScaleNanos))
+                .build();
+    }
+
+    public static EmaSpec ofTime(@NotNull final String timestampCol, long timeScaleNanos) {
+        return ImmutableEmaSpec.builder()
+                .control(EmaControl.DEFAULT)
+                .timeScale(TimeScale.ofTime(timestampCol, timeScaleNanos))
+                .build();
     }
 
     public static EmaSpec ofTicks(@NotNull EmaControl control, long tickWindow) {
-        return ImmutableEmaSpec.of(control, null, tickWindow);
+        return ImmutableEmaSpec.builder()
+                .control(control)
+                .timeScale(TimeScale.ofTicks(tickWindow))
+                .build();
     }
 
-    @Parameter
-    public abstract EmaControl control();
+    public static EmaSpec ofTicks(long tickWindow) {
+        return ImmutableEmaSpec.builder()
+                .control(EmaControl.DEFAULT)
+                .timeScale(TimeScale.ofTicks(tickWindow))
+                .build();
+    }
 
     @Parameter
-    @Nullable
-    public abstract String timestampCol();
+    public abstract EmaControl control();
 
     @Parameter
-    public abstract long timeScaleUnits();
+    public abstract TimeScale timeScale();
```

Comment:
Everything in the hierarchy of an immutable should (ideally) be immutable itself. I see that TimeScale is not - and does not implement hashcode, equals, etc. 

I think Timescale should be an interface w/ visitor pattern, and two impls TimescaleTicks, TimescaleDuration.

Reply:
This is used internally, never exposed to the user.  Is there a need for the complexity of a class hierarchy?